File: cs50_Assignments/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    long cardnumber = 0;
    int z = 0;
    int l = 0;
    do
    {
        cardnumber = get_long("Input your credit card number : ");
    }
    while (cardnumber < 0);
    int i = 0;
    long number = cardnumber;
    do
    {
        cardnumber = (cardnumber / 10);
        i++;
    }
    while (cardnumber > 0);
    for (int j = 0; j < i; j++)
    {

        if (j % 2 == 0)
        {
            long e = pow(10, j);
            int x = (number / (10 * e)) % 10;
            int y = (number / e) % 10;
            if (2 * x > 9)
            {
                z = z + (((2 * x) / 10) + ((2 * x) % 10));
            }
            else
            {
                z = z + 2 * x;
            }
            l = l + y;
        }
    }
    int credit = z + l;
    long d = pow(10, (i - 1));
    long f = pow(10, (i - 2));
    long g = pow(10, (i - 4));
    if (credit % 10 != 0)
    {
        printf("INVALID\n");
    }
    else if (number / d == 4 && (i == 13 || i == 16))
    {
        printf("VISA\n");
    }
    else if ((number / f > 50 && number / f < 56 && i == 16) || (number / g > 2220 && number / g < 2721&& i == 16))
    {
        printf("MASTERCARD\n");
    }
    else if ((number / f == 34 || number / f == 37) && i == 15)
    {
        printf("AMEX\n");
    }
    else
    {
      printf("INVALID\n");
    }
}

File: week6/sentimental-credit/credit.py
import re


def main():
    card_number = input("Enter Credit-Card Number :")
    card_check(card_number)
    length_card = len(card_number)
    pattern2 = re.compile("^5[1-5]")
    pattern3 = re.compile("^3[4,7]")
    if card_number[0] == '4' and (length_card == 13 or length_card == 16):
        print("VISA")
    elif pattern2.match(card_number) and length_card == 16:
        print("MASTERCARD")
    elif pattern3.match(card_number) and length_card == 15:
        print("AMEX")
    else:
        print("INVALID")
        return 3


def card_check(card_number):

    pattern = re.compile("^[0-9]{13}|^[0-9]{15}|^[0-9]{16}$")
    while True:
        if pattern.search(card_number) == None:
            print("INVALID")
            return 1
        else:
            break
    check_for_validity(card_number)


def check_for_validity(card):
    sum1 = 0
    sum2 = 0
    length = len(card)
    for i in range(length-2, -1, -2):
        cardx2 = int(card[i])*2
        if cardx2 > 9:
            sum1 = sum1 + cardx2 % 10 + 1
        else:
            sum1 = sum1 + cardx2
    for i in range(length-1, -1, -2):
        sum2 = sum2 + int(card[i])
    if (sum1 + sum2) % 10 != 0:
        print("INVALID")
        return 2


main()
